iot_config/src/gateway_service.rs

```diff
@@ -1,17 +1,130 @@
+}
+
+impl GatewayService {
+    pub async fn new(settings: &Settings) -> Result<Self> {
+        let pool = settings.database.connect(10).await?;
+        let region_map = RegionMap::new(&pool).await?;
+        Ok(Self {
+            admin_pubkey: settings.admin_pubkey()?,
+            follower_service: FollowerService::from_settings(&settings.follower),
+            pool,
+            region_map,
+            signing_key: settings.signing_keypair()?,
+        })
+    }
 
-pub struct GatewayService {}
+    fn verify_admin_signature<R>(&self, request: R) -> Result<R, Status>
+    where
+        R: MsgVerify,
+    {
+        request
+            .verify(&self.admin_pubkey)
+            .map_err(|_| Status::permission_denied("invalid admin signature"))?;
+        Ok(request)
+    }
+}
 
 #[tonic::async_trait]
 impl iot_config::Gateway for GatewayService {
     async fn region_params(
         &self,
-        _request: Request<GatewayRegionParamsReqV1>,
+        request: Request<GatewayRegionParamsReqV1>,
     ) -> GrpcResult<GatewayRegionParamsResV1> {
-        unimplemented!()
+        let request = request.into_inner();
+
+        let pubkey = PublicKey::try_from(request.address.clone())
+            .map_err(|_| Status::invalid_argument("invalid gateway address"))?;
+        request
+            .verify(&pubkey)
+            .map_err(|_| Status::permission_denied("invalid request signature"))?;
+
+        let GatewayInfo { location, gain, .. } = self
+            .follower_service
+            .clone()
+            .resolve_gateway_info(&pubkey.into())
+            .await
+            .map_err(|_| Status::internal("gateway lookup error"))?;
+
+        let location = {
+            let location =
+                location.ok_or_else(|| Status::internal("gateway location undefined"))?;
+            Cell::from_raw(location)
+                .map_err(|_| Status::internal("gateway location is not a valid h3 index"))?
+        };
+
+        let region = self
+            .region_map
+            .get_region(location)
+            .await
+            .ok_or_else(|| Status::internal("region lookup failed for asserted location"))?;
+        let params = self.region_map.get_params(&region).await;
+
+        let mut resp = GatewayRegionParamsResV1 {
+            region: region.into(),
+            params,
+            gain: gain as u64,
+            signature: vec![],
+        };
+        resp.signature = self
+            .signing_key
+            .sign(&resp.encode_to_vec())
+            .map_err(|_| Status::internal("resp signing error"))?;
+        Ok(Response::new(resp))
+    }
+
+    async fn load_region(&self, request: Request<LoadRegionReqV1>) -> GrpcResult<LoadRegionResV1> {
+        let request = request.into_inner();
+        let req = self.verify_admin_signature(request)?;
+
+        let region = Region::from_i32(req.region)
+            .ok_or_else(|| Status::invalid_argument("invalid region"))?;
+
+        let params = match req.params {
```

Comment:
re the match on Some/None, wont the proto itself prevent a none in this instance ?  the proto field is not optional

Reply:
still don't understand exactly how this works but the compiler requires this to be an option even though it's not listed as optional in the proto

Reply:
as i recall, it has something to do with the combination of the fact that all values in proto3 are "technically" optional and how proto3 treats default values for non-scalar types, of which region would be one

Reply:
right.. good recall and a common cause of confusion